tests/xdc/history_replication_conflict_test.go

```diff
@@ -0,0 +1,389 @@
+func (s *historyReplicationConflictTestSuite) executeNamespaceReplicationTasksUntil(ctx context.Context, operation enumsspb.NamespaceOperation, version int64) {
+	for {
+		task := <-s.namespaceReplicationTasks
+		err := s.namespaceTaskExecutor.Execute(ctx, task)
+		s.NoError(err)
+		if task.NamespaceOperation == operation && task.FailoverVersion == version {
+			return
+		}
+	}
+}
+
+// executeHistoryReplicationTasksUntil executes buffered history event replication tasks until the specified event type
+// is encountered for the workflowId.
+func (s *historyReplicationConflictTestSuite) executeHistoryReplicationTasksUntil(
+	ctx context.Context,
+	cluster string,
+	eventType enumspb.EventType,
+) {
+	for {
+		select {
+		case t := <-s.historyReplicationTasks[cluster]:
+			events := s.executeHistoryReplicationTask(t)
+			for _, event := range events {
+				if event.GetEventType() == eventType {
+					return
+				}
+			}
+		case <-ctx.Done():
+			s.FailNow("timed out waiting for replication task to be processed")
+		}
+	}
+}
+
+func (s *historyReplicationConflictTestSuite) executeHistoryReplicationTask(task *hrcTestExecutableTask) []*historypb.HistoryEvent {
+	serializer := serialization.NewSerializer()
+	trackableTask := (*task).TrackableExecutableTask
+	err := trackableTask.Execute()
+	s.NoError(err)
+	task.result <- err
+	attrs := (*task).replicationTask.GetHistoryTaskAttributes()
+	s.NotNil(attrs)
+	s.Equal(s.workflowId, attrs.WorkflowId)
+	events, err := serializer.DeserializeEvents(attrs.Events)
+	s.NoError(err)
+	return events
+}
+
+func (c *hrcTestNamespaceReplicationTaskExecutor) Execute(ctx context.Context, task *replicationspb.NamespaceTaskAttributes) error {
+	c.s.namespaceReplicationTasks <- task
+	// Report success, although we have merely buffered the task and will execute it later.
+	return nil
+}
+
+// Convert the replication tasks using the base converter, and wrap them in our own executable tasks.
+func (t *hrcTestExecutableTaskConverter) Convert(
+	taskClusterName string,
+	clientShardKey replication.ClusterShardKey,
+	serverShardKey replication.ClusterShardKey,
+	replicationTasks ...*replicationspb.ReplicationTask,
+) []replication.TrackableExecutableTask {
+	convertedTasks := t.converter.Convert(taskClusterName, clientShardKey, serverShardKey, replicationTasks...)
+	testExecutableTasks := make([]replication.TrackableExecutableTask, len(convertedTasks))
+	for i, task := range convertedTasks {
+		testExecutableTasks[i] = &hrcTestExecutableTask{
+			taskClusterName:         taskClusterName,
+			s:                       t.s,
+			TrackableExecutableTask: task,
+			replicationTask:         replicationTasks[i],
+			result:                  make(chan error),
+		}
+	}
+	return testExecutableTasks
+}
+
+// Execute pushes the task to a buffer and waits for it to be executed.
+func (t *hrcTestExecutableTask) Execute() error {
+	t.s.historyReplicationTasks[t.taskClusterName] <- t
+	return <-t.result
```

Comment:
A possible concern here is that the task could time out, if the test does not execute it promptly. My response to that concern is
1. As written, I believe it's safe
2. The alternative was significantly less clean -- I found I had to return a retryable error, and the set a retry policy that in practice never retries during the test, and even then found that I was having to special-case the first history event replication task (the one that replicates the events on workflow start).